src/coreclr/jit/optimizer.cpp

```diff
@@ -7608,17 +7608,805 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+        {
+            // Given the following sequence of blocks :
+            //        B1: brtrue(t1, B3)
+            //        B2: brtrue(t2, BX)
+            //        B3:
+            // we will try to fold it to :
+            //        B1: brtrue((!t1)&&t2, BX)
+            //        B3:
+
+            this->sameTarget = false;
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+    else
+    {
+        // Does b1 jump to b3?
+        // One example: Given the following sequence of blocks :
+        //        B1: brtrue(!t1, B3)
+        //        B2: return(t2)
+        //        B3: return(false)
+        //    we will try to fold it to :
+        //        B1: return(t1|t2)
+        if (b1->bbJumpDest != b2->bbNext)
+        {
+            return nullptr;
+        }
+    }
+
+    // Find the block conditions of b1 and b2
+
+    if (b2->countOfInEdges() > 1 || (optReturnBlock && b3->countOfInEdges() > 1))
+    {
+        return nullptr;
+    }
+
+    // Find the condition for the first block
+
+    Statement* s1 = b1->lastStmt();
+
+    GenTree* testTree1 = s1->GetRootNode();
+    noway_assert(testTree1->gtOper == GT_JTRUE);
+
+    // The second and the third block must contain a single statement
+
+    Statement* s2 = b2->firstStmt();
+    if (s2->GetPrevStmt() != s2)
+    {
+        return nullptr;
+    }
+
+    GenTree* testTree2 = s2->GetRootNode();
+
+    if (!optReturnBlock)
+    {
+        noway_assert(testTree2->gtOper == GT_JTRUE);
+    }
+    else
+    {
+        if (testTree2->gtOper != GT_RETURN)
+        {
+            return nullptr;
+        }
+
+        Statement* s3 = b3->firstStmt();
+        if (s3->GetPrevStmt() != s3)
+        {
+            return nullptr;
+        }
+
+        GenTree* testTree3 = s3->GetRootNode();
+        if (testTree3->gtOper != GT_RETURN)
+        {
+            return nullptr;
+        }
+
+        if (!varTypeIsIntegral(testTree2->TypeGet()) || !varTypeIsIntegral(testTree3->TypeGet()))
+        {
+            return nullptr;
+        }
+
+        // The third block is Return with "CNS_INT int 0/1"
+        if (testTree3->AsOp()->gtOp1->gtOper != GT_CNS_INT)
+        {
+            return nullptr;
+        }
+
+        if (testTree3->AsOp()->gtOp1->gtType != TYP_INT)
+        {
+            return nullptr;
+        }
+
+        this->t3 = testTree3;
+    }
+
+    this->t1.testTree = testTree1;
+    this->t2.testTree = testTree2;
+
+    return s1;
+}
+
+//
+// optOptimizeBoolsChkTypeCostCond: Checks if type conditions meet the folding condition, and
+//                                  if cost to fold is not too expensive
+//
+// Arguments:
+//      comp    Instance of Compiler class
+//
+// Return:
+//      True if it meets type conditions and cost conditions.	Else false.
+//
+bool OptBoolsDsc::optOptimizeBoolsChkTypeCostCond(Compiler* comp)
+{
+    noway_assert(t1.compTree->OperIs(GT_EQ, GT_NE) && t1.compTree->AsOp()->gtOp1 == c1);
```

Comment:
Do these really need to be `noway_assert`, or would `assert` be sufficient?

Reply:
Done.